### PR TITLE
[JENKINS-36240] Better implementation of getTrustedRevision

### PR DESCRIPTION
**Pull request by jglick**
Downstream of https://github.com/jenkinsci/github-api-plugin/pull/11.

@reviewbybees

### CODE REVIEW

**Comment by jglick** [comment]
> we need to wait for the API to be confirmed before we can merge this

Probably. Or at least figure out how to fall back to earlier behavior in the meantime.

> we may need to retain the old behaviour for tokens that do not have permission to do the permission check

AFAIK those would not have had permission to use the old API either.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java [hunk 1]
```diff
@@ -843,69 +825,47 @@ private void checkout(GitSCM scm, Run<?,?> build, GitClient git, TaskListener li
...
-                        collaboratorNames = Collections.singleton(repoOwner);
+                    String permission = repo.getPermission(sourceOwner).getPermission();
+                    listener.getLogger().format("Permission check of %s on %s/%s says: %s%n", sourceOwner, repoOwner, repository, permission);
+                    trusted = permission.equals("admin") || permission.equals("write");
+                } catch (FileNotFoundException x) {
+                    listener.getLogger().format("Failed to do a direct permission check of %s on %s/%s; scan token has no access to this repository, or endpoint removed%n", sourceOwner, repoOwner, repository);
+                    trusted = false; // TODO JENKINS-37608: make this configurable
+                } catch (HttpException x) {
+                    if (x.getResponseCode() == 403) {
+                        listener.getLogger().format("Not authorized to do a direct permission check of %s on %s/%s; scan token has no access to this repository%n", sourceOwner, repoOwner, repository);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java]: > Cache this so you don't retry the request for all PRs

Not sure why caching would be required; this code should only be run during an actual build of the branch project. We could _try_ to cache 403s per repository × `credentialsId`, though this cache could become stale for several reasons—different token, perhaps different permissions for a given token, different repository or team settings…

**Comment by jglick** [comment]
> wait for the API to be confirmed before we can merge this

Or just implement JENKINS-37608 so that the user can configure a fixed behavior for all pull requests. The current implementation is so flawed I see little point in retaining it.

**Comment by jglick** [comment]
I think what I want this to do is:

* trust nonforked PRs unconditionally, as it already does
* for a forked PR,
   * try using the experimental API, and accept the result if it comes in
   * else check a configuration option (at repo or org level) about whether to trust all forked PRs, or none
   * else, if not configured, do not trust it

**Comment by stephenc** [comment]
The functionality of this change will be reviewed after https://issues.jenkins-ci.org/browse/JENKINS-43426 has been delivered

**Comment by jglick** [comment]
Indeed this will likely need to be rewritten to use the proposed new API.

**Comment by jglick** [comment]
Needs to be reworked to offer an alternative to `ForkPullRequestDiscoveryTrait.TrustContributors` based on the GitHub API which is now official at least on github.com (no idea about GHE status): https://github.com/kohsuke/github-api/pull/358

**Comment by stephenc** [comment]
@jglick did you see #148?

**Pull request closed without merging**